### PR TITLE
Google OSS-Fuzz integration

### PR DESCRIPTION
**Pull request by vy**
As a deliverable of #2891 and #2892, implements Google OSS-Fuzz integration. See the added `FUZZING.adoc` for details.

### CODE REVIEW

File: log4j-slf4j2-impl-fuzz-test/src/main/java/org/apache/logging/slf4j/fuzz/Slf4jLoggerFacade.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+    @Override
+    public void log(String message, Throwable throwable) {
+        logger.error(message, throwable);
+    }
+
+    @Override
+    public void log(String message, Object[] parameters) {
+        logger.error(message, parameters);
+    }
```
Comment (ppkarwasz) [inline log4j-slf4j2-impl-fuzz-test/src/main/java/org/apache/logging/slf4j/fuzz/Slf4jLoggerFacade.java:52]:
```suggestion
    private static final String FQCN = Slf4jLoggerFacade.class.getName();

    private final Logger logger;

    private Slf4jLoggerFacade(final Logger logger) {
        this.logger = logger;
    }

    static Slf4jLoggerFacade ofClass(final Class<?> clazz) {
        requireNonNull(clazz, "clazz");
        final Logger logger = LoggerFactory.getLogger(clazz);
        return new Slf4jLoggerFacade(logger);
    }

    private LoggingEventBuilder atError() {
        LoggingEventBuilder builder = logger.atError();
        if (builder instanceof CallerBoundaryAware) {
            ((CallerBoundaryAware) builder).setCallerBoundary(FQCN);
        }
        return builder;
    }

    @Override
    public void log(final String message) {
        atError().log(message);
    }

    @Override
    public void log(String message, Throwable throwable) {
        atError().setCause(throwable).log(message);
    }

    @Override
    public void log(String message, Object[] parameters) {
        atError().log(message, parameters);
    }
```

**Review by ppkarwasz — CHANGES_REQUESTED** [review]
Great job! :100:

I only have some minor remarks on this huge PR:

**Comment by vy** [comment]
@ppkarwasz, what do your FQCN fixes address in the context of fuzzing?

File: log4j-fuzz-test/src/main/java/org/apache/logging/log4j/fuzz/FuzzingUtil.java [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+        // Create the conclusive configuration (uninitialized!)
+        final Configuration config = configBuilder
+                .setStatusLevel(Level.OFF)
+                .add(appenderComponentBuilder)
+                .add(loggerComponentBuilder)
+                .build(false);
+
+        // Initialize the configuration
+        return Configurator.initialize(config);
```
[the remark this answers is not in the document]
Comment (vy) [reply log4j-fuzz-test/src/main/java/org/apache/logging/log4j/fuzz/FuzzingUtil.java]: Fixed in d8622d9ce6bfcf256bc55a987fb5381afc65a437.

File: log4j-core-fuzz-test/src/main/java/org/apache/logging/log4j/core/fuzz/PatternLayoutFuzzer.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+                        .newLayout("PatternLayout")
+                        // Enforce using a single message-based converter, i.e., `MessagePatternConverter`
+                        .addAttribute("pattern", "%m"));
+        final Logger logger = loggerContext.getLogger(PatternLayoutFuzzer.class);
+        return new Log4jLoggerFacade(logger);
+    }
+
+    public static void fuzzerTestOneInput(final FuzzedDataProvider dataProvider) {
+        fuzzLogger(LOGGER, dataProvider);
+    }
```
[the remark this answers is not in the document]
Comment (vy) [reply log4j-core-fuzz-test/src/main/java/org/apache/logging/log4j/core/fuzz/PatternLayoutFuzzer.java]: Fixed in d8622d9ce6bfcf256bc55a987fb5381afc65a437.

File: log4j-layout-template-json-fuzz-test/src/main/java/org/apache/logging/log4j/layout/template/json/fuzz/JsonTemplateLayoutFuzzer.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    private static LoggerFacade createLogger() {
+        final LoggerContext loggerContext = createLoggerContext(
+                JsonEncodingAppender.PLUGIN_NAME, configBuilder -> configBuilder.newLayout("JsonTemplateLayout"));
+        final Logger logger = loggerContext.getLogger(JsonTemplateLayoutFuzzer.class);
+        return new Log4jLoggerFacade(logger);
+    }
+
+    public static void fuzzerTestOneInput(final FuzzedDataProvider dataProvider) {
+        fuzzLogger(LOGGER, dataProvider);
+    }
```
[the remark this answers is not in the document]
Comment (vy) [reply log4j-layout-template-json-fuzz-test/src/main/java/org/apache/logging/log4j/layout/template/json/fuzz/JsonTemplateLayoutFuzzer.java]: Fixed in d8622d9ce6bfcf256bc55a987fb5381afc65a437.

**Review by ppkarwasz — APPROVED** [review]

**Pull request merged**